File: python/zpy/src/zingg_py/features.py
"""Feature vectors — port of ``Features.scala``.

Adds a ``z_features`` column: per pair, the per-field/per-matcher similarity
scores concatenated into a numpy vector (the Spark ML dense vector in the
original).
"""

from __future__ import annotations

import numpy as np
import pandas as pd

from . import similarity
from .schema import ZinggConf

FEATURE_COL = "z_features"


def feature_total(cfg: ZinggConf) -> int:
    return sum(similarity.feature_width(mt) for f in cfg.fields for mt in f.match_types)


def _row_features(row, cfg: ZinggConf) -> np.ndarray:
    arr: list[float] = []
    for f in cfg.fields:
        a = row.get(f"{ZinggConf.LEFT_PREFIX}{f.name}")
        b = row.get(f"{ZinggConf.RIGHT_PREFIX}{f.name}")
        for mt in f.match_types:
            arr.extend(similarity.features(mt, a, b))
    return np.asarray(arr, dtype=float)


def add_features(pairs: pd.DataFrame, cfg: ZinggConf) -> pd.DataFrame:
    out = pairs.copy()
    if pairs.empty:
        out[FEATURE_COL] = pd.Series([], dtype=object)
        return out
    feats = [_row_features(row, cfg) for _, row in pairs.iterrows()]
    out[FEATURE_COL] = pd.Series(feats, index=pairs.index)
    return out


def feature_matrix(df: pd.DataFrame) -> np.ndarray:
    """Stack the ``z_features`` column into an ``n x d`` matrix."""
    if df.empty:
        return np.zeros((0, 0))
    return np.vstack(df[FEATURE_COL].to_numpy())

File: zscala/src/main/scala/zingg/zs/Features.scala
package zingg.zs

import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.sql.{DataFrame, Row}
import org.apache.spark.sql.functions.{col, struct, udf}

object Features {

  val FeatureCol = "z_features"

  def addFeatures(pairs: DataFrame, cfg: ZinggConf): DataFrame = {
    val fields = cfg.fields
    val widths = fields.map(f => Similarity.featureWidth(f.matchType))
    val total  = widths.sum
    val featUdf = udf { (row: Row) =>
      val arr = new Array[Double](total)
      var idx = 0
      fields.foreach { f =>
        val a = row.getAs[Any](s"${ZinggConf.LeftPrefix}${f.name}")
        val b = row.getAs[Any](s"${ZinggConf.RightPrefix}${f.name}")
        val v = Similarity.features(f.matchType, a, b)
        System.arraycopy(v, 0, arr, idx, v.length)
        idx += v.length
      }
      Vectors.dense(arr)
    }
    val structCols = pairs.columns.map(col)
    pairs.withColumn(FeatureCol, featUdf(struct(structCols: _*)))
  }
}
